File: erlang/pe019.erl
-module(pe019).
-export([start/0]).

january() -> 31.
february() -> 28.
march() -> 31.
april() -> 30.
may() -> 31.
june() -> 30.
july() -> 31.
august() -> 31.
september() -> 30.
october() -> 31.
november() -> 30.
december() -> 31.
leap_february() -> 29.

num_cycles() -> 25.

year() -> [january(), february(), march(), april(), may(), june(), july(), august(), september(), october(), november(), december()].
leap_year() -> [january(), leap_february(), march(), april(), may(), june(), july(), august(), september(), october(), november(), december()].
cycle() -> year() ++ year() ++ year() ++ leap_year(). % inefficient but probably preferable to next?
%cycle() -> [
%  january(), february(), march(), april(), may(), june(), july(), august(), september(), october(), november(), december(),
%  january(), february(), march(), april(), may(), june(), july(), august(), september(), october(), november(), december(),
%  january(), february(), march(), april(), may(), june(), july(), august(), september(), october(), november(), december(),
%  january(), leap_february(), march(), april(), may(), june(), july(), august(), september(), october(), november(), december()].

start() ->
  pe019(num_cycles(), cycle(), 2, 0).
  
pe019(NC, _C, _DN, DC) when NC == 0 ->
  io:format("~w~n", [DC]);
  
pe019(NC, C, DN, DC) ->
  {TN, TC} = process_month(C, DN, DC),
  pe019(NC - 1, C, TN, TC).

process_month([], DN, DC) ->
  {DN, DC};

process_month([H|T], DN, DC) ->
  Day = (DN + H) rem 7,
  if
    Day == 0 ->
      process_month(T, Day, DC + 1);
    true ->
      process_month(T, Day, DC)
  end.

File: ocaml/pe019.ml
open Core.Std

let january       = 31 ;;
let february      = 28 ;;
let march         = 31 ;;
let april         = 30 ;;
let may           = 31 ;;
let june          = 30 ;;
let july          = 31 ;;
let august        = 31 ;;
let september     = 30 ;;
let october       = 31 ;;
let november      = 30 ;;
let december      = 31 ;;
let leap_february = 29 ;;

let num_cycles    = 25 ;;

let mardec    = [march; april; may; june; july; august; september; october; november; december] ;;
let year      = List.append [january; february] mardec ;;
let leap_year = List.append [january; leap_february] mardec ;;
let cycle     = List.append (List.append year year) (List.append year leap_year) ;;

let rec process_month months daynum daycount =
  match months with
  | [] -> (daynum, daycount)
  | head::body ->
    let day = ((daynum + head) mod 7) in 
      if day = 0 then
        process_month body day (daycount + 1)
      else
        process_month body day daycount ;;

let rec pe019 count months daynum daycount =
  if count = 0 then
    daycount
  else
    let (tn, tc) = (process_month months daynum daycount) in
      pe019 (count - 1) months tn tc ;;

printf "%d\n" (pe019 num_cycles cycle 2 0) ;;
